### PR TITLE
CIRC-440: prevent multiple recalls from changing the loan due date

### PR DESCRIPTION
**Pull request by mreno-EBSCO**
The loan due date should only be changed once due to a recall. Subsequent recalls should not modify the due date. We will now set and respect the `dueDateChangedByRecall` flag. It will be set on a recall, if it has not already been set, and when it comes time to apply recall logic to update the loan, if this flag is set, we no longer perform the recall logic.

This fix also has changed an expectation of CIRC-289 (moving recall requests) and the unit test was updated to reflect the new expected outcome.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
  mod-circulation-storage
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/circulation/domain/Loan.java [hunk 1]
```diff
@@ -104,6 +105,12 @@ public Loan changeDueDate(DateTime newDueDate) {
     return this;
   }
 
+  public Loan changeDueDateChangedByRecall(boolean newDueDateChangedByRecall) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/Loan.java]: As I understand it, there is only need to mark this as having happened, and not reset it. Hence, is there any need for a parameter here?

File: src/main/java/org/folio/circulation/domain/Loan.java [hunk 2]
```diff
@@ -174,6 +181,10 @@ public boolean isOpen() {
     return StringUtils.equals(getStatus(), "Open");
   }
 
+  public boolean isDueDateChangedByRecall() {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/Loan.java]:
```suggestion
  public boolean hasDueDateChangedDueToRecall() {
```

```suggestion
  public boolean wasDueDateChangedByRecall() {
```

File: src/main/java/org/folio/circulation/domain/UpdateLoan.java [hunk 3]
```diff
@@ -74,6 +66,24 @@ public UpdateLoan(Clients clients,
     return of(() -> loanAndRelatedRecords);
   }
 
+  private CompletableFuture<Result<RequestAndRelatedRecords>> doRecall(Loan loan,
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/UpdateLoan.java]:
```suggestion
  private CompletableFuture<Result<RequestAndRelatedRecords>> recall(Loan loan,
```

File: src/test/java/api/requests/scenarios/MoveRequestTests.java [hunk 4]
```diff
@@ -932,8 +945,8 @@ public void canUpdateSourceAndDestinationLoanDueDateOnMoveRecallRequest() throws
         .equals(interestingTimesItem.getId().toString()))
       .findFirst().orElse(new JsonObject());
 
-    assertThat(smallAngryPlanetLoan.getInstant("dueDate"), is(expectedLoanDueDate));
-    assertThat(interestingTimesLoan.getInstant("dueDate"), is(expectedLoanDueDate));
+    assertThat(smallAngryPlanetLoan.getInstant("dueDate"), is(expectedJamesLoanDueDate));
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/scenarios/MoveRequestTests.java]: Why are the due dates set to the system date and time? Is that because the policy being used has a zero recall return interval?

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java [hunk 5]
```diff
@@ -525,6 +526,400 @@ public void changedDueDateAfterRecallingAnItemShouldRespectTenantTimezone()
...
+        .unlimitedRenewals()
+        .renewFromSystemDate()
+        .withRecallsMinimumGuaranteedLoanPeriod(Period.weeks(2))
+        .withRecallsRecallReturnInterval(Period.weeks(1));
+
+    final IndividualResource loanPolicy = loanPoliciesFixture.create(canCirculateRollingPolicy);
+
+    useLoanPolicyAsFallback(loanPolicy.getId(),
+        requestPoliciesFixture.allowAllRequestPolicy().getId(),
+        noticePoliciesFixture.activeNotice().getId());
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Do notices need to be active for this test?

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java [hunk 6]
```diff
@@ -525,6 +526,400 @@ public void changedDueDateAfterRecallingAnItemShouldRespectTenantTimezone()
...
+
+    final IndividualResource loan = loansFixture.checkOutByBarcode(
+      smallAngryPlanet, jessica, ClockManager.getClockManager().getDateTime());
+
+    final String originalDueDate = loan.getJson().getString("dueDate");
+
+    JsonObject storedLoan = loansStorageClient.getById(loan.getId()).getJson();
+
+    final String recalledDueDate = storedLoan.getString("dueDate");
+    assertThat("due date after recall is not the original date",
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: This check appears to be that the recalled due date _is_ the same as the original due date?

I'm finding it hard to follow what should have happened here. 

Should the due date of the loan already have been reduced, due to a recall having been placed before the check out? If so, an additional recall request shouldn't change it further.

Or is it that pre-existing recall requests aren't accounted for during the initial check out? Only when a new recall is placed after the check out?

File: src/main/java/org/folio/circulation/domain/Loan.java (hunk 1 above)
Comment (mreno-EBSCO) [reply src/main/java/org/folio/circulation/domain/Loan.java]: When I was coding it initially, there was a question as to whether or not this should be reset if the loan is renewed after a cancelled recall request is made. You are correct that Cate has indicated that this is not possible now. I'll remove the parameter.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java [hunk 7]
```diff
@@ -525,6 +526,400 @@ public void changedDueDateAfterRecallingAnItemShouldRespectTenantTimezone()
...
+    requestsFixture.placeHoldShelfRequest(smallAngryPlanet, charlotte,
+        DateTime.now(DateTimeZone.UTC), requestServicePoint.getId(), "Recall");
+
+    storedLoan = loansStorageClient.getById(loan.getId()).getJson();
+    assertThat("second recall changed the due date (2 weeks)",
+        storedLoan.getString("dueDate"), is(recalledDueDate));
+  }
+
+  @Test
+  public void pagedItemRecalledThenLoanedBecomesOverdueAndNextRecallDoesNotChangeDueDate()
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Why is becoming overdue relevant to this logic?

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java [hunk 8]
```diff
@@ -525,6 +526,400 @@ public void changedDueDateAfterRecallingAnItemShouldRespectTenantTimezone()
...
+    requestsFixture.placeHoldShelfRequest(smallAngryPlanet, charlotte,
+        DateTime.now(DateTimeZone.UTC), requestServicePoint.getId(), "Recall");
+
+    storedLoan = loansStorageClient.getById(loan.getId()).getJson();
+    assertThat("second recall changed the due date (2 weeks)",
+        storedLoan.getString("dueDate"), is(recalledDueDate));
+  }
+
+  @Test
+  public void secondRecallRequestWithMGDTruncationInPlaceDoesNotChangeDueDate()
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Given that the minimum duration is sooner than the recall interval after the system date, the due date shouldn't have changed even before this change was made. Is that correct?

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java [hunk 9]
```diff
@@ -525,6 +526,400 @@ public void changedDueDateAfterRecallingAnItemShouldRespectTenantTimezone()
...
+        DateTime.now(DateTimeZone.UTC), requestServicePoint.getId(), "Recall");
+
+    JsonObject storedLoan = loansStorageClient.getById(loan.getId()).getJson();
+
+    final String recalledDueDate = storedLoan.getString("dueDate");
+    assertThat("due date is the original date",
+        recalledDueDate, not(originalDueDate));
+
+    final String expectedDueDate = ClockManager.getClockManager().getDateTime().plusWeeks(2).toString(ISODateTimeFormat.dateTime());
+    assertThat("due date is not the recall due date (2 weeks)",
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: The check suggests the due date should be the minimum loan period, the message states that it is not.

Am I just getting confused reading the assertion messages, because they state what shouldn't have happened, rather than what should?

File: src/test/java/api/requests/scenarios/MoveRequestTests.java (hunk 4 above)
Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/MoveRequestTests.java]: This was an existing test. I assume the default loan policy being applied to the tests doesn't have a minimum guaranteed due date duration, so we apply the recall with the system date here, which is the default behavior. In the recall tests I wrote I supplied a loan policy with explicit intervals.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java [hunk 10]
```diff
@@ -525,6 +526,400 @@ public void changedDueDateAfterRecallingAnItemShouldRespectTenantTimezone()
...
+    requestsFixture.placeHoldShelfRequest(smallAngryPlanet, charlotte,
+        DateTime.now(DateTimeZone.UTC), requestServicePoint.getId(), "Recall");
+
+    storedLoan = loansStorageClient.getById(loan.getId()).getJson();
+    assertThat("second recall changed the due date (2 weeks)",
+        storedLoan.getString("dueDate"), is(recalledDueDate));
+  }
+
+  @Test
+  public void secondRecallRequestWithRDTruncationInPlaceDoesNotChangeDueDate()
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Is this attempting to demonstrate that when the recall return interval is applied, it isn't applied for the second recall?

If so, it might be possible to simplify this by not setting a minimum guaranteed period, rather than artificially setting the recall return interval to more than the guaranteed period.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java [hunk 11]
```diff
@@ -525,6 +526,400 @@ public void changedDueDateAfterRecallingAnItemShouldRespectTenantTimezone()
...
+    requestsFixture.placeHoldShelfRequest(smallAngryPlanet, charlotte,
+        DateTime.now(DateTimeZone.UTC), requestServicePoint.getId(), "Recall");
+
+    storedLoan = loansStorageClient.getById(loan.getId()).getJson();
+    assertThat("second recall changed the due date (2 months)",
+        storedLoan.getString("dueDate"), is(recalledDueDate));
+  }
+
+  @Test
+  public void itemRecalledThenCancelledAndNextRecallDoesNotChangeDueDate()
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Thank you for adding this scenario explicitly

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 5 above)
Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: That's a good question. I know very little about notices or how they work in the tests. Should I use `noticePoliciesFixture.inactiveNotice().getId()`? This code was added when CIRC-203 was implemented.

File: src/test/java/api/requests/scenarios/MoveRequestTests.java (hunk 4 above)
Comment (marcjohnson-kint) [reply src/test/java/api/requests/scenarios/MoveRequestTests.java]: I understand it is an existing test and does not fall fully within the scope of this work. I only noticed it due to the changes. 

Maybe it is out of scope of this change to improve this test?

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 6 above)
Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Hmm... I think this assertion is probably a copy/paste error. I don't see the need for it at all. I'll get rid of it. The important assertion is after the second recall.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 7 above)
Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: It probably isn't, but it was specified in the scenarios list in CIRC-440, so I wrote a test for it, just to be safe.

Comment (marcjohnson-kint) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Ok, that makes sense, if it is important to the process owner, it seems reasonable to explicitly test it

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 5 above)
Comment (marcjohnson-kint) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Yes please. I've been starting to do that in places where polices are set and the test doesn't involve notices.

I've been wondering if there should be something stateful that knows about the current policies and rules, so can instead only change the policy the test cares about.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 9 above)
Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: I think that is the problem here. I agree that it is confusing in the test itself. I need to rethink this or do some more research into what best practices are for unit test failure messages. Should we describe the assertion or the failure of the assertion. In the log you'd see something like "something.something.SomeAssertionError: due date is not the recall due date (2 weeks)". Is that clearer than "something.something.SomeAssertionError: due date is the recall due date (2 weeks)"?

Again, maybe it is better to use a less specific message so that the logs look something like "something.something.SomeAssertionError: due date should be the recall due date (2 weeks)". Or we just drop the messages and let it log something like "something.something.SomeAssertionError: dueDate (some date) not equal to expected expectedDueDate (some other date)" or whatever the default message is.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 8 above)
Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: I'm not sure I follow. In this test the recall interval plus the system date (7 days) is sooner than the minimum guaranteed duration (14 days), so we apply the MGD to the original loan when the recall is placed. This test shows subsequent recalls do not affect the due date.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 10 above)
Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: We could, though that is a slightly different code path (I believe the second statement in an "or" instead of the first), but I am pretty sure the tests already cover that path. I'll try it out (it should work just fine).

File: src/test/java/api/requests/scenarios/MoveRequestTests.java (hunk 4 above)
Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/MoveRequestTests.java]: Do you think we need explicit policies set here? I didn't think there was anything wrong with the test, I just needed to make it work since the expectations had changed.

Comment (marcjohnson-kint) [reply src/test/java/api/requests/scenarios/MoveRequestTests.java]: I found it hard to follow why the due date was meant to be the system time. (And maybe the policy that if no recall return interval is set, it expects immediate return, which seems odd to me in general).

I mostly just didn’t understand, let’s defer this to another time

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 5 above)
Comment (mtraneis) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Could the code to set up loan policy be extracted out for reuse and to de-clutter the tests, if it's the same for every test?

Comment (mreno-EBSCO) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: The loan policy itself is not the same in every test, though there is commonality between several of them. Maybe we can do this as follow on work? Perhaps some of this could be moved to the APITests base class so tests in other domains could use policies that have minimum guaranteed durations and recall intervals set for tests that involve recalls.

Comment (mtraneis) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Sure, a follow on work is fine. This is not a big deal.  You're right, the loan policy itself is not the same. By that I meant the useLoanPolicyAsFallback() call.

Comment (marcjohnson-kint) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: The configuration of circulation rules in the tests needs significant work.

The current approach of using the API test base methods was a compromise that allowed many of the tests to have a sensible default and the ability to change it relatively easily.

I'd prefer some form of stateful actor that looked after the rules.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 9 above)
Comment (marcjohnson-kint) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: The biggest limitation I've found with the default diagnostic information is that it only has knowledge of the values, not where the values come from (in this case the due date of the loan).

I think I probably write quite vague messages fairly often, e.g. `due date should have changed` or `due date should not have changed`.

File: src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java (hunk 8 above)
Comment (marcjohnson-kint) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: Yes, you are right, I got the two things the wrong way round.

What I was trying to say was, if both recalls are placed within the minimum duration, then the due date wouldn't have changed even with the current code

Comment (marcjohnson-kint) [reply src/test/java/api/requests/scenarios/LoanDueDatesAfterRecallTests.java]: And I've just noticed the offset, so I'll close this

(the details of the clock freezing really needs hiding away, but that is for another day)
